src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs

```diff
@@ -154,10 +154,29 @@ public static DateTimeOffset GetLastWriteTime(string fullPath)
             return data.ftLastWriteTime.ToDateTimeOffset();
         }
 
-        private static void MoveDirectory(string sourceFullPath, string destFullPath, bool sameDirectoryDifferentCase)
+        internal static void MoveDirectory(string sourceFullPath, string destFullPath)
         {
+            // Source and destination must have the same root.
+            ReadOnlySpan<char> sourceRoot = Path.GetPathRoot(sourceFullPath);
+            ReadOnlySpan<char> destinationRoot = Path.GetPathRoot(destFullPath);
+            if (!sourceRoot.Equals(destinationRoot, StringComparison.OrdinalIgnoreCase))
+            {
+                throw new IOException(SR.IO_SourceDestMustHaveSameRoot);
+            }
+
             if (!Interop.Kernel32.MoveFile(sourceFullPath, destFullPath, overwrite: false))
             {
+                // Source and destination must not be the same file.
+                // We don't check upfront, but handle this as a MoveFile failure.
+                // MoveFile will fail because it requires the source to exist and the destination to not exist.
+                // That is not possible if they are the same file.
+                ReadOnlySpan<char> srcNoDirectorySeparator = Path.TrimEndingDirectorySeparator(sourceFullPath.AsSpan());
+                ReadOnlySpan<char> destNoDirectorySeparator = Path.TrimEndingDirectorySeparator(destFullPath.AsSpan());
+                if (srcNoDirectorySeparator.Equals(destNoDirectorySeparator, PathInternal.StringComparison))
```

Comment:
The `sameDirectoryDifferentCase` stuff is confusing to me. I think this is what it was meant to check.